### PR TITLE
collapse large diffs && collapse/expand any diff within multi file patch view

### PR DESCRIPTION
**Pull request by annthurium**
### Description of the Change

This pull request adds code to detect large diffs, and automatically collapse them. It also adds the ability for users to collapse and expand any diff within a multi file patch view.  There's a lot of shared logic between both so it made sense to tackle them at the same time.

### TODO

https://github.com/atom/github/projects/18

### Screenshot/Gif

TBD

### Benefits

- The pull request view, commit preview, and commit pane are more useful now.  
- No more endless scrolling to get past your package-lock.json changes that you probably don't need to review in such detail!

### Possible Drawbacks

TBD

### Applicable Issues

https://github.com/atom/github/issues/1860

### QA Checklist
- [x] file patches that exceed large diff threshold are initially collapsed
- [x] file patches that do not exceed large diff threshold are not initially collapsed
- [x] collapsing by clicking "collapse" icon collapses the file patch and all decorations:
   - [x] comments
   - [x] hunk headers
   - [x] file mode headers
   - [x] symlink headers
 - [x] collapsing by clicking in the diff collapses the file patch and all decorations:
   - [x] comments (only exist in the pull request files changed tab)
   - [x] hunk headers
   - [x] file mode headers
   - [x] symlink headers
 - [x] expanding a previously collapsed patch re-expands the patch and all decorations are present
- [x] when collapsing and expanding, patches stay in the correct order
- [x] test collapsing / expanding if only one patch exists
- [ ] ~selected hunks should still be selected if new files are added to the patch~ (see https://github.com/atom/github/projects/18#card-18069394)
- [ ] ~selected lines should still be selected if new files are added to the patch~ (see https://github.com/atom/github/projects/18#card-18069394)
- [x] staging view should update when new files are added
- [x] no visual regressions in the component trees that use multi file patch:
    - [x] pull request changed files tab
    - [x] commit detail
    - [x] commit preview
    - [x] changed files 

### Metrics

- Count of how many files are manually collapsed / expanded
- Performance of rendering large diffs

### Tests

TBD

### Documentation

TBD

### Release Notes

The GitHub package now supports collapsing and expanding any files within multi file diff views (such as commit preview and the pull request files view), and also automatically collapses large files.

### User Experience Research (Optional)

The UI is simple enough for this that I would argue it does not warrant UXR.

### CODE REVIEW

**Comment by smashwilson** [comment]
We have basic collapse and expand :tada:

collapse-expand

It's a little janky. And I'm not sure where the extra newline is coming from. But it's a start!

**Comment by 50Wliu** [comment]
:wave: I've been using this branch and just hit this deprecation:
marker-deprecation
I'm on 79fd7dbde43aa9b35126715c0736ddde4a034950.

**Comment by simurai** [comment]
Nice to have:

Be able to collapse/uncollapse **all** files. On .com you can use the <kbd>alt</kbd> key when clicking.

screen shot 2019-01-30 at 9 54 36 am

Hard to discover, but doesn't need an extra toggle.

File: lib/models/patch/builder.js [hunk 1]
```diff
@@ -1,31 +1,43 @@
...
 import {Unchanged, Addition, Deletion, NoNewline} from './region';
 import FilePatch from './file-patch';
 import MultiFilePatch from './multi-file-patch';
 
-export function buildFilePatch(diffs) {
-  const layeredBuffer = initializeBuffer();
+export const DEFAULT_OPTIONS = {
+  // Number of lines after which we consider the diff "large"
+  // TODO: Set this based on performance measurements
+  largeDiffThreshold: 100,
```
[the remark this answers is not in the document]
Comment (Aerijo) [reply lib/models/patch/builder.js]: @annthurium Ah, right. My point was that arbitrary choices that aren't configurable tend to have issues raised about them, and cloning the entire package to make one setting configurable is not worth the headache.

I would suggest adding a config option, either explicitly in settings or hidden with a default and documented in the README. Metrics are also an option, though I don't know if they'd be particularly useful in this case (I don't think many people would tweak this setting).

On a related note, would it be possible to measure diff size using bytes instead? What of a (pathological) scenario where they edit a minified, single line file?

File: docs/react-component-atlas.md [hunk 2]
```diff
@@ -74,8 +74,8 @@ This is a high-level overview of the structure of the React component tree that
 > > > [`<MultiFilePatchController>`](/lib/controllers/multi-file-patch-controller.js)
 > > > [`<MultiFilePatchView>`](/lib/views/multi-file-patch-view.js)
 > > >
-> > > Render a sequence of git-generated file patches within a TextEditor, using decorations to include contextually
-> > > relevant controls.
+> > > Render a sequence of git-generated file patches within a TextEditor, using decorations to include contextually relevant controls.
+> > > See [`MultiFilePatchView` atlas](#multifilepatchview-atlas) below for a more detailed breakdown.
```
Comment (smashwilson) [inline docs/react-component-atlas.md]: What I _really_ want is a tool that parses our source and outputs this atlas as an .svg diagram or something 🤔

File: lib/atom/atom-text-editor.js [hunk 3]
```diff
@@ -81,15 +86,21 @@ export default class AtomTextEditor extends React.Component {
...
+      }
+
       return null;
     });
   }
 
   componentDidUpdate(prevProps) {
     const modelProps = extractProps(this.props, editorUpdateProps);
     this.getRefModel().map(editor => editor.update(modelProps));
+    this.observeEmptiness();
```
[the remark this answers is not in the document]
Comment (smashwilson) [reply lib/atom/atom-text-editor.js]: I can't seem to accept this suggestion, and I'm not sure why?

File: lib/containers/pr-changed-files-container.js [hunk 4]
```diff
@@ -99,6 +106,13 @@ export default class PullRequestChangedFilesContainer extends React.Component {
       if (response && response.ok) {
         const rawDiff = await response.text();
         const multiFilePatch = this.buildPatch(rawDiff);
+
+        this.mfpSubs.dispose();
+        this.mfpSubs = new CompositeDisposable();
+        for (const fp of multiFilePatch.getFilePatches()) {
+          this.mfpSubs.add(fp.onDidChangeRenderStatus(() => this.forceUpdate()));
```
Comment (smashwilson) [inline lib/containers/pr-changed-files-container.js]: These `.forceUpdate()` calls are kind of a smell to me. I don't think we can do better right now without some serious restructuring of our state management, though.

File: lib/models/patch/builder.js [hunk 5]
```diff
@@ -1,31 +1,49 @@
...
+import Patch, {TOO_LARGE, EXPANDED} from './patch';
 import {Unchanged, Addition, Deletion, NoNewline} from './region';
 import FilePatch from './file-patch';
 import MultiFilePatch from './multi-file-patch';
 
-export function buildFilePatch(diffs) {
-  const layeredBuffer = initializeBuffer();
+export const DEFAULT_OPTIONS = {
+  // Number of lines after which we consider the diff "large"
+  // TODO: Set this based on performance measurements
```
Comment (smashwilson) [inline lib/models/patch/builder.js]: 🚨 🔥 this obsolete "TODO"

File: lib/models/patch/multi-file-patch.js [hunk 6]
```diff
@@ -44,45 +44,41 @@ export default class MultiFilePatch {
...
-        addition: this.getAdditionLayer(),
-        deletion: this.getDeletionLayer(),
-        noNewline: this.getNoNewlineLayer(),
-      },
+      patchBuffer: opts.patchBuffer !== undefined ? opts.patchBuffer : this.getLayeredBuffer(),
       filePatches: opts.filePatches !== undefined ? opts.filePatches : this.getFilePatches(),
     });
   }
 
+  getLayeredBuffer() {
```
Comment (smashwilson) [inline lib/models/patch/multi-file-patch.js]: 🚨 I should really rename this to `getPatchBuffer()` for consistency. `getLayeredBuffer()` is a relic of the pre-PatchBuffer model.

File: lib/models/patch/multi-file-patch.js [hunk 7]
```diff
@@ -209,71 +204,30 @@ export default class MultiFilePatch {
...
-      layers: {
-        patch: buffer.addMarkerLayer(),
-        hunk: buffer.addMarkerLayer(),
-        unchanged: buffer.addMarkerLayer(),
-        addition: buffer.addMarkerLayer(),
-        deletion: buffer.addMarkerLayer(),
-        noNewline: buffer.addMarkerLayer(),
-      },
-    };
+    this.patchBuffer = nextPatchBuffer;
```
Comment (smashwilson) [inline lib/models/patch/multi-file-patch.js]: So much less messy now :smile:

File: lib/views/file-patch-header-view.js [hunk 8]
```diff
@@ -25,9 +26,15 @@ export default class FilePatchHeaderView extends React.Component {
     undoLastDiscard: PropTypes.func.isRequired,
     diveIntoMirrorPatch: PropTypes.func.isRequired,
     openFile: PropTypes.func.isRequired,
+    // should probably change 'toggleFile' to 'toggleFileStagingStatus'
+    // because the addition of another toggling function makes the old name confusing.
```
Comment (smashwilson) [inline lib/views/file-patch-header-view.js]: Honestly I've never been happy with the "toggle" function names. None of the alternatives I've been able to come up with are succinct enough though. We need a nice, single verb that means "stage or unstage"...

File: lib/views/multi-file-patch-view.js [hunk 9]
```diff
@@ -97,9 +102,65 @@ export default class MultiFilePatchView extends React.Component {
         this.props.refInitialFocus && this.props.refInitialFocus.setter(element);
       }),
     );
+
+    // Synchronously maintain the editor's scroll position and logical selection across buffer updates.
+    this.suppressChanges = false;
+    let lastScrollTop = null;
+    let lastScrollLeft = null;
+    let lastSelectionIndex = null;
+    if (this.props.onWillUpdatePatch && this.props.onDidUpdatePatch) {
```
Comment (smashwilson) [inline lib/views/multi-file-patch-view.js]: We could do this more cleanly by specifying default props for these subscription methods. (For example, if this component is rendered with `onWillUpdatePatch` but not `onDidUpdatePatch`, neither will be called, which feels weird.)

File: lib/views/pr-review-comments-view.js [hunk 10]
```diff
@@ -11,6 +11,7 @@ import GithubDotcomMarkdown from './github-dotcom-markdown';
 import Timeago from './timeago';
 
 export default class PullRequestCommentsView extends React.Component {
+  // do we even need the relay props here? does not look like they are used
```
Comment (smashwilson) [inline lib/views/pr-review-comments-view.js]: 🚨 ☝️ Let's clean this up before we merge.

File: styles/atom-text-editor.less [hunk 11]
```diff
@@ -3,3 +3,8 @@
   width: 100%;
   height: 100%;
 }
+
+// Conceal the automatically added blank line element within TextEditors that we want to actually be empty
```
[the remark this answers is not in the document]
Comment (smashwilson) [reply styles/atom-text-editor.less]: We aren't adding blank lines. If you render a TextEditor with an empty buffer, the TextEditor will render a single line element anyway, so you can start typing. But it looks wrong in a patch view that consists of only decorations, so I just hid it with CSS.

File: test/builder/pr.js [hunk 12]
```diff
@@ -6,16 +6,22 @@ class CommentBuilder {
...
     this._replyTo = null;
+    this._isMinimized = false;
   }
 
   id(i) {
     this._id = i;
     return this;
   }
 
+  minmized(m) {
```
Comment (smashwilson) [inline test/builder/pr.js]: 🚨 Spelling!

File: test/models/patch/file-patch.test.js [hunk 13]
```diff
@@ -308,9 +316,7 @@ describe('FilePatch', function() {
     let unstageLayeredBuffer;
 
     beforeEach(function() {
-      const buffer = new TextBuffer();
-      const layers = buildLayers(buffer);
-      unstageLayeredBuffer = {buffer, layers};
+      unstageLayeredBuffer = new PatchBuffer();
```
Comment (smashwilson) [inline test/models/patch/file-patch.test.js]: We should probably rename this too :eyes:

File: test/models/patch/multi-file-patch.test.js [hunk 14]
```diff
@@ -776,4 +781,478 @@ describe('MultiFilePatch', function() {
...
+
+          multiFilePatch.expandFilePatch(fp0);
+          assertInFilePatch(fp0, multiFilePatch.getBuffer()).hunks(hunk({index: 0, start: 0}));
+          assertInFilePatch(fp1, multiFilePatch.getBuffer()).hunks(hunk({index: 1, start: 4}));
+          assertInFilePatch(fp2, multiFilePatch.getBuffer()).hunks(hunk({index: 2, start: 8}));
+          assertInFilePatch(fp3, multiFilePatch.getBuffer()).hunks(hunk({index: 3, start: 12, last: true}));
+        });
+      });
+
+      it('is deterministic regardless of the order in which collapse and expand operations are performed', function() {
```
Comment (smashwilson) [inline test/models/patch/multi-file-patch.test.js]: I'm still proud of this test case 😆

File: test/models/patch/patch-buffer.test.js [hunk 15]
```diff
@@ -0,0 +1,438 @@
...
+    patchBuffer.markPosition('hunk', [0, 1]);
+
+    patchBuffer.clearAllLayers();
+
+    assert.lengthOf(patchBuffer.findMarkers('patch', {}), 0);
+    assert.lengthOf(patchBuffer.findMarkers('hunk', {}), 0);
+  });
+
+  describe('extractPatchBuffer', function() {
+    it('extracts a subset of the buffer and layers as a new LayeredBuffer', function() {
```
Comment (smashwilson) [inline test/models/patch/patch-buffer.test.js]: 🚨 Missed the rename here 👀

File: test/models/patch/patch.test.js [hunk 16]
```diff
@@ -165,9 +166,7 @@ describe('Patch', function() {
     let stageLayeredBuffer;
 
     beforeEach(function() {
-      const stageBuffer = new TextBuffer();
-      const stageLayers = buildLayers(stageBuffer);
-      stageLayeredBuffer = {buffer: stageBuffer, layers: stageLayers};
+      stageLayeredBuffer = new PatchBuffer();
```
Comment (smashwilson) [inline test/models/patch/patch.test.js]: 🚨 Rename, rename

File: test/models/patch/patch.test.js [hunk 17]
```diff
@@ -372,9 +367,7 @@ describe('Patch', function() {
     let unstageLayeredBuffer;
 
     beforeEach(function() {
-      const unstageBuffer = new TextBuffer();
-      const unstageLayers = buildLayers(unstageBuffer);
-      unstageLayeredBuffer = {buffer: unstageBuffer, layers: unstageLayers};
+      unstageLayeredBuffer = new PatchBuffer();
```
Comment (smashwilson) [inline test/models/patch/patch.test.js]: 🚨 Rename...

File: lib/atom/atom-text-editor.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (smashwilson) [reply lib/atom/atom-text-editor.js]: Added manually :wink:

File: docs/react-component-atlas.md [hunk 18]
```diff
@@ -74,8 +74,8 @@ This is a high-level overview of the structure of the React component tree that
 > > > [`<MultiFilePatchController>`](/lib/controllers/multi-file-patch-controller.js)
 > > > [`<MultiFilePatchView>`](/lib/views/multi-file-patch-view.js)
 > > >
-> > > Render a sequence of git-generated file patches within a TextEditor, using decorations to include contextually
-> > > relevant controls.
```
Comment (kuychaco) [inline docs/react-component-atlas.md]: for testing: comment on deleted line

File: docs/react-component-atlas.md [hunk 19]
```diff
@@ -178,3 +178,36 @@ This is a high-level overview of the structure of the React component tree that
 > > > [`<GithubTileView>`](/lib/views/changed-files-count-view.js)
 > > >
 > > > Displays the GitHub logo. Clicking it opens the GitHub tab.
+
+
+
+## `MultiFilePatchView` Atlas
```
Comment (kuychaco) [inline docs/react-component-atlas.md]: another test comment

File: docs/react-component-atlas.md (hunk 18 above)
Comment (kuychaco) [reply docs/react-component-atlas.md]: does this work?

**Pull request merged**